rclrs/src/node/builder.rs

```diff
@@ -40,6 +44,10 @@ pub struct NodeBuilder {
     context: Arc<Mutex<rcl_context_t>>,
     name: String,
     namespace: String,
+    domain_id: usize,
+    use_global_arguments: bool,
+    arguments: Vec<String>,
+    enable_rosout: bool,
```

Comment:
Thanks for your suggestions! I have one question. If making fields public, should context filed also be public? Current context field has raw type of rcl and should not expose those types against users as much as possible I think. If it should make context field private, it is better for other fields to make private because of the rule, `Either make all fields of a type public, or make none of them public`, from blog post that is linked from reddit you attached.  Another way to make all field public is usage of reference for Context, but this may make compiler more sensitive and cause lifetime errors which is hard to understand why, although I haven't inspect yet.

By the way, surely, current builder setters are smell ones if I do not misunderstand the means about smell( I read posts as smell setter is equal to trivial setter, and trivial setter is the setter that takes a parameter and only assigns to struct field). However I could not understand why setter is worse when there are smell setters. If there are comments says why smell setters are bad, please tell me.

Also, the post you attached does not says about assigning a value to public field as far as I know. I read that post as recommendation of using C-like struct instantiation pattern with Default when the struct just be data box. However, on another note, this pattern is possible only if there are only optional values (there is also same questionable comment about that but no response), and NodeBuilder has essential fields (i.e. `context` and `name`), so, in this case, I cannot apply this pattern to NodeBuilder.

Reply:
Sorry, I guess that link is a bit confusing without context. Let me try to clarify
- I agree, the context should definitely be private.
- The "trivial setters" matklad calls "smelly" are those in the parent comment: `shape.set_position(Vec3(0, 9, 2));` etc. – they have a signature like `fn set_position(&mut self, position: Vec3)`.
  - This signature is different from the chainable setters in `NodeBuilder`, which look more like `fn position(self, position: Vec3) -> Self`. Chainable setters are useful and I think no one sees them as a code smell. So I think there's nothing wrong at all with the current implementation of `NodeBuilder`.
- So, the point of linking to that comment was just that public fields are not always bad. The "builder lite" blog post it's attached to isn't really relevant here. Sorry for the confusion.
- Further down in the reddit thread, people agree that public fields are fine unless (1) there are some invariants that need to be protected (we don't have that, since validation happens later) or (2) you need to be careful about making breaking changes to the struct (which is also not a concern, since we can make breaking changes, and since the private context field will make the changes non-breaking anyway).
- Yep, https://matklad.github.io/2022/05/29/binary-privacy.html recommends categorizing everything into data or abstract data types. Not sure which one a builder is supposed to be. I think it's closer to data, since we could also remove the context field and only pass it in in the `build()` method, then everything could be public. But in the end it's only a rule of thumb, just like "avoid trivial setters".

It's a really small change we're discussing here, so I'll leave it up to you :) I just find it nice to see the fields in the docs and see that there is nothing weird going on, and to be able to also use `builder.enable_rosout = true;` syntax.

Reply:
Feel free to resolve this, it was just a suggestion and like I said I'll leave it up to you.